### PR TITLE
[OpenMP][libomptarget] Enable automatic unified shared memory executi…

### PR DESCRIPTION
**Pull request by carlobertolli**
…on (zero-copy) on MI300A.

This patch enables applications that did not request OpenMP unified_shared_memory to run with the same zero-copy behavior, where mapped memory does not result in extra memory allocations and memory copies, but CPU-allocated memory is accessed from the device. The name for this behavior is "automatic zero-copy" and it relies on detecting: that the runtime is running on a MI300A, that the user did not select unified_shared_memory in their program, and that XNACK (unified memory support) is enabled in the current GPU configuration. If all these conditions are met, then automatic zero-copy is triggered.

This patch is still missing support for global variables, which will be provided in a subsequent patch.

### CODE REVIEW

File: openmp/libomptarget/src/PluginManager.cpp [hunk 1]
```diff
@@ -146,6 +146,7 @@ void PluginAdaptorTy::initDevices(PluginManager &PM) {
   ExclusiveDevicesAccessor->reserve(DeviceOffset + NumPD);
   for (int32_t PDevI = 0, UserDevId = DeviceOffset; PDevI < NumPD; PDevI++) {
     auto Device = std::make_unique<DeviceTy>(this, UserDevId, PDevI);
+
```
Comment (shiltian) [inline openmp/libomptarget/src/PluginManager.cpp:149]: Unrelated

File: openmp/libomptarget/src/OpenMP/Mapping.cpp [hunk 2]
```diff
@@ -252,8 +252,9 @@ TargetPointerResultTy MappingInfoTy::getTargetPointer(
       MESSAGE("device mapping required by 'present' map type modifier does not "
               "exist for host address " DPxMOD " (%" PRId64 " bytes)",
               DPxPTR(HstPtrBegin), Size);
-  } else if (PM->getRequirements() & OMP_REQ_UNIFIED_SHARED_MEMORY &&
-             !HasCloseModifier) {
+  } else if ((PM->getRequirements() & OMP_REQ_UNIFIED_SHARED_MEMORY &&
+              !HasCloseModifier) ||
+             PM->getUseAutoZeroCopy()) {
```
Comment (shiltian) [inline openmp/libomptarget/src/OpenMP/Mapping.cpp:257]: The format looks off here

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
+
```
Comment (shiltian) [inline openmp/libomptarget/test/mapping/auto_zero_copy.cpp:1]: empty leading line

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+// RUN: %libomptarget-compilexx-generic
+// RUN: env HSA_XNACK=1 LIBOMPTARGET_DEBUG=1 %libomptarget-run-generic 2>&1 \
+// RUN: | %fcheck-generic -check-prefix=DEBUG -check-prefix=CHECK
+// REQUIRES: libomptarget-debug
+
+// UNSUPPORTED: aarch64-unknown-linux-gnu
```
Comment (shiltian) [inline openmp/libomptarget/test/mapping/auto_zero_copy.cpp:7]: Instead of listing all architectures that are not supported, probably better to just require AMD

File: openmp/libomptarget/include/PluginManager.h [hunk 5]
```diff
@@ -190,6 +190,11 @@ struct PluginManager {
   /// Add \p Flags to the user provided requirements.
   void addRequirements(int64_t Flags) { Requirements.addRequirements(Flags); }
 
+  bool getUseAutoZeroCopy() const { return UseAutoZeroCopy; }
```
Comment (shiltian) [inline openmp/libomptarget/include/PluginManager.h:193]: Just out of curiosity, why is it called auto zero copy instead of zero copy?

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 6]
```diff
@@ -2913,7 +2913,7 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
   /// Create an AMDGPU plugin and initialize the AMDGPU driver.
   AMDGPUPluginTy()
       : GenericPluginTy(getTripleArch()), Initialized(false),
-        HostDevice(nullptr) {}
+        HostDevice(nullptr), IsAPU(false), OMPX_HSAXnack("HSA_XNACK", false) {}
```
Comment (shiltian) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:2916]: TBH I think all these class members can be initialized directly when the class is defined. No need to list them here.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 7]
```diff
@@ -2990,6 +2990,8 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
     if (auto Err = HostDevice->init())
       return std::move(Err);
 
+    isAPU();
```
Comment (shiltian) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:2993]:
```suggestion
    checkIfAPU();
```

Comment (shiltian) [reply openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:2993]: `isAPU` doesn't conform with LLVM code standard.

File: openmp/libomptarget/include/Shared/PluginAPI.h [hunk 8]
```diff
@@ -225,6 +225,7 @@ int32_t __tgt_rtl_initialize_record_replay(int32_t DeviceId, int64_t MemorySize,
                                            void *VAddr, bool isRecord,
                                            bool SaveOutput,
                                            uint64_t &ReqPtrArgOffset);
+int32_t __tgt_rtl_use_auto_zero_copy();
```
Comment (jdoerfert) [inline openmp/libomptarget/include/Shared/PluginAPI.h:228]: Docs.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 9]
```diff
@@ -3109,6 +3126,47 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
...
+
+    std::string StrGfxName(GfxName);
+    std::transform(std::begin(StrGfxName), std::end(StrGfxName),
+                   std::begin(StrGfxName),
+                   [](char c) { return std::tolower(c); });
+
+    if (StrGfxName == "gfx940") {
+      IsAPU = true;
+      return;
+    } else if (StrGfxName == "gfx942") {
```
Comment (shiltian) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3154]: Early return doesn't need `else if`.

File: openmp/libomptarget/include/PluginManager.h [hunk 10]
```diff
@@ -208,6 +213,13 @@ struct PluginManager {
 
   /// Devices associated with plugins, accesses to the container are exclusive.
   ProtectedObj<DeviceContainerTy> Devices;
+
+  /// Whe running on an APU, the GPU plugin may decide to
+  /// run in zero-copy even though the user did not program
+  /// their application with unified_shared_memory requirement.
+  /// This execution modality is called auto zero-copy and it is
+  /// cached information from the Plugin.
+  bool UseAutoZeroCopy = false;
```
Comment (jdoerfert) [inline openmp/libomptarget/include/PluginManager.h:222]: I'd figure this is a device property. I can easily imagine an APU system with a dedicated card and this property should not apply to all devices. If we make it global, it doesn't need a flag, we could simply implicitly set OMP_REQ_UNIFIED_SHARED_MEMORY in the requirements, no?

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp (hunk 7 above)
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:2993]: Docs.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp (hunk 6 above)
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:2916]: (I wrote this before I saw Shilei's comment)
I understand you followed existing code but I think we should (maybe in a pre-commit) move these initializations to the member definition. It's cumbersome to have them here, in order, as one needs to look at 2 places to understand type and initial value.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 11]
```diff
@@ -3109,6 +3126,47 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
...
 
+  /// Detect if current architecture is an APU.
+  void isAPU() {
+    char GfxName[64];
+
+    if (!KernelAgents.size())
+      return;
+
+    // Do not allow for mixed APU+discrete GPU combinations: an APU can only be
+    // that, so only check the first GPU agent.
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3137]: I don't understand why we assume you can't have an MI250X and a MI300A in the same system.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 12]
```diff
@@ -3109,6 +3126,47 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
...
+    hsa_agent_t GPUAgent = KernelAgents[0];
+    std::memset((void *)&GfxName, 0, sizeof(char) * 64);
+
+    hsa_status_t Status = hsa_agent_get_info(
+        GPUAgent, (hsa_agent_info_t)HSA_AGENT_INFO_NAME, GfxName);
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3142]: This is already available in ComputeUnitKind

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 13]
```diff
@@ -3109,6 +3126,47 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
...
+    if (Status != HSA_STATUS_SUCCESS)
+      return;
+
+    std::string StrGfxName(GfxName);
+    std::transform(std::begin(StrGfxName), std::end(StrGfxName),
+                   std::begin(StrGfxName),
+                   [](char c) { return std::tolower(c); });
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3149]: If we assume the case might not be uniform, let's do the transform when we initialize ComputeUnitKind.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 14]
```diff
@@ -3122,6 +3180,12 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
 
   /// The device representing all HSA host agents.
   AMDHostDeviceTy *HostDevice;
+
+  /// Is the plugin associated with an APU?
+  bool IsAPU;
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3185]: Again, this sounds to me like a device property.

File: openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h [hunk 15]
```diff
@@ -88,6 +88,34 @@ StringRef parseTargetID(StringRef TargetID, StringMap<bool> &FeatureMap) {
   return Arch;
 }
 
+bool IsXnackEnabledViaKernelParam() {
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h:91]: Coding style, both name and non-inline definitions in headers.

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp (hunk 4 above)
Comment (jdoerfert) [reply openmp/libomptarget/test/mapping/auto_zero_copy.cpp:7]: Does the require look for the target arch or if we can compile for the target?

**Review by jdoerfert — COMMENTED** [review]
I like that we get auto-magic for these things.

High-level questions:

1) Why is this not a device property?
2) Why is this not just an implicit OMP_REQ_UNIFIED_SHARED_MEMORY?

**Comment by carlobertolli** [comment]
> I like that we get auto-magic for these things.
> 
> High-level questions:
> 
> 1. Why is this not a device property?
> 2. Why is this not just an implicit OMP_REQ_UNIFIED_SHARED_MEMORY?

Good questions.
For 1. I am not expecting an APU to have any PCI-e (or other mean) attached GPUs, at this point in time.
For 2. We need to distinguish the handling of global variables (not yet supported by this patch). When an application is built with requires usm, global variables in the device binary become references to the original host globals. This doesn't happen when not building with requires unified_shared_memory, hence the need to differentiate between "automatic" zero-copy and user-requested zero-copy.
In auto zero-copy, global variables need to be copied between their host and device items, following the map(always,.. constructs requested by the user (remember, the user did not use requires usm, so they are required to use map clauses for everything, including global variables).

Thanks for the rest of the style comments, really appreciated. I will apply the changes tomorrow.

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp (hunk 4 above)
Comment (shiltian) [reply openmp/libomptarget/test/mapping/auto_zero_copy.cpp:7]: Oh right. It just checks if we can compile. We will still need them here.

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp [hunk 16]
```diff
@@ -0,0 +1,43 @@
+
+// RUN: %libomptarget-compilexx-generic
+// RUN: env HSA_XNACK=1 LIBOMPTARGET_DEBUG=1 %libomptarget-run-generic 2>&1 \
+// RUN: | %fcheck-generic -check-prefix=DEBUG -check-prefix=CHECK
+// REQUIRES: libomptarget-debug
```
Comment (jplehr) [inline openmp/libomptarget/test/mapping/auto_zero_copy.cpp:5]: I wonder if we should prefer to use `LIBOMPTARGET_INFO` with the appropriate level to check for data copies. This is mostly motivated as we run buildbots with pure release builds w/o support for `LIBOMPTARGET_DEBUG`.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 17]
```diff
@@ -3122,6 +3180,12 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
...
+
+  /// Value of HSA_XNACK environment variable.
+  BoolEnvar OMPX_HSAXnack;
```
Comment (jplehr) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3188]: Why is this `OMPX`? I think those names are sort-of reserved for OMP extensions, right?

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 18]
```diff
@@ -3065,6 +3067,21 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
...
+    // Automatic zero-copy only applies to programs that did
+    // not request unified_shared_memory and are deployed on an
+    // APU with XNACK enabled.
+    if (getRequiresFlags() & OMP_REQ_UNIFIED_SHARED_MEMORY)
+      return false;
+
+    // XNACK can be enabled with with kernel boot parameter or with
+    // environment variable.
+    return (IsAPU &&
+            (utils::IsXnackEnabledViaKernelParam() || OMPX_HSAXnack.get()));
```
Comment (jplehr) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3082]: No need for `.get()` call here, `BoolEnvar` is convertible to bool.

File: openmp/libomptarget/include/PluginManager.h (hunk 5 above)
Comment (carlobertolli) [reply openmp/libomptarget/include/PluginManager.h:193]: zero-copy (or no-copy) is what is commonly called the support to requires unified_shared_memory. In that case, the user explicitly asks for the support, as opposed to the proposed support which turns on "automatically" if the system is an APU and XNACK (unified memory) support is enabled.

File: openmp/libomptarget/include/PluginManager.h (hunk 10 above)
Comment (carlobertolli) [reply openmp/libomptarget/include/PluginManager.h:222]: Fair enough: I do not have such a card to test against, but it is conceivable that someone will eventually have it. What I am thinking is a check on every device in the system to see if they are all an APU and would turn auto zero-copy on, then we turn it on in libomptarget. If one or more of the devices is not an APU, then it will return false to the query and we will not do auto zero-copy.
I think this should cover any future arch and we can think about mixing and matching based on the device num that is passed to getTargetPointer and getTgtPtr in a later patch.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp (hunk 6 above)
Comment (carlobertolli) [reply openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:2916]: actually, the existing code had the initialization in the correct place. This is plainly my mistake for this patch.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp (hunk 11 above)
Comment (carlobertolli) [reply openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3137]: Replied above, will fix.

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp (hunk 16 above)
Comment (carlobertolli) [reply openmp/libomptarget/test/mapping/auto_zero_copy.cpp:5]: I checked LIBOMPTARGET_INFO and it does not print anything related to this. If you think it makes sense, I can add an info print so we can check that the mapping of the arrays is done through auto zero-copy.

Comment (jdoerfert) [reply openmp/libomptarget/test/mapping/auto_zero_copy.cpp:5]: INFO should print if there is a data copy, no? So the absence of that is what we could test for. In that case I'd like a negative test too though.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp (hunk 17 above)
Comment (jdoerfert) [reply openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:3188]: I think we use things like `LIBOMPTARGET_AMDGPU_TEAMS_PER_CU`, but this is really piggy backing on the HSA env var. We will later need XXXXEnvar with multiple names, for now, OMPX_HSAXnack seems ok to me.

File: openmp/libomptarget/include/PluginManager.h (hunk 10 above)
Comment (jdoerfert) [reply openmp/libomptarget/include/PluginManager.h:222]: I'm fine with doing it per plugin for now, but I would like to avoid the extra bool here and the getter/setter. Can we add a new requirements flag:
`OMPX_REQ_AUTO_ZERO_COPY`
and reuse the requirement stuff instead of adding a on-off bool in many places?

Comment (carlobertolli) [reply openmp/libomptarget/include/PluginManager.h:222]: I am happy to have a new OMP_REQ_AUTO_ZERO_COPY requirement flag.
I am already done moving the property to a device one and I would have just to change this to set the req flag instead of having the boolean check. Will do that today if this makes sense.

File: openmp/libomptarget/src/PluginManager.cpp [hunk 19]
```diff
@@ -144,19 +144,33 @@ void PluginAdaptorTy::initDevices(PluginManager &PM) {
...
+    UseAutoZeroCopy = false;
   for (int32_t PDevI = 0, UserDevId = DeviceOffset; PDevI < NumPD; PDevI++) {
     auto Device = std::make_unique<DeviceTy>(this, UserDevId, PDevI);
+
     if (auto Err = Device->init()) {
       DP("Skip plugin known device %d: %s\n", PDevI,
          toString(std::move(Err)).c_str());
       continue;
     }
+    UseAutoZeroCopy = Device->useAutoZeroCopy() && UseAutoZeroCopy;
```
Comment (jdoerfert) [inline openmp/libomptarget/src/PluginManager.cpp:160]: The other way around to save calls into the plugin.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp [hunk 20]
```diff
@@ -2618,6 +2623,15 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
     return Plugin::success();
   }
 
+  /// Returns true if auto zero-copy the best configuration for the current
+  /// arch.
+  bool useAutoZeroCopyImpl() override {
+    // XNACK can be enabled with with kernel boot parameter or with
+    // environment variable.
+    return (IsAPU &&
+            (utils::isXnackEnabledViaKernelParam() || HSAXnackEnv.get()));
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:2632]: I would swap the latter two. The EnvVar is cheaper to check.

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp [hunk 21]
```diff
@@ -0,0 +1,43 @@
+
+// RUN: %libomptarget-compilexx-generic
+// RUN: env HSA_XNACK=1 LIBOMPTARGET_INFO=30 %libomptarget-run-generic 2>&1 \
+// RUN: | %fcheck-generic -check-prefix=INFO -check-prefix=CHECK
+// REQUIRES: libomptarget-debug
```
Comment (jdoerfert) [inline openmp/libomptarget/test/mapping/auto_zero_copy.cpp:5]: not needed anymore.

**Review by jdoerfert — COMMENTED** [review]
I think this addressed all but one of my main concerns.
Last thing is the testing.
We should add tests for the different interactions (e.g., verify non-HSA_XNACK will copy, so will omp requires).
Further, what happens if this is executed on a non-APU system? Won't it fail?

**Comment by carlobertolli** [comment]
> I think this addressed all but one of my main concerns. Last thing is the testing. We should add tests for the different interactions (e.g., verify non-HSA_XNACK will copy, so will omp requires). Further, what happens if this is executed on a non-APU system? Won't it fail?

I will add checks in the current test for when auto zero-copy is not triggered and we need to see a copy, and a test for unified_shared_memory, where we want to see the same log as in zero-copy.

For non APU systems, this is my thinking:
- If the GPU supports unified memory (think amdgpu's with xnack support, such as MI200, or nvidia GPUs with unified meomory support), then we want to offer auto zero-copy to the users, but in this case it would be hidden behind an environment variable (e.g., OMPX_AUTO_ZERO_COPY or OMPX_APU_MAPS). The reasoning is: on discrete GPU systems, the best default is to make device memory allocations and h2d/d2h copies. The user needs to do something special (set the env variable, turn on xnack) to trigger auto zero-copy, because we expect it to result, on average, in worse performance.
- If the GPU does not support unified memory, then auto zero-copy won't be triggered automatically. If the user sets OMPX_AUTO_ZERO_COPY then I would imagine it would fail at the first CPU memory access from a GPU.

I will work on better diagnostic of all cases in a subsequent patch. Example: build application with unified_shared_memory and run it on a system with unified memory disabled should result in a warning that things might not work and a suggestion to turn on unified memory support. This applies to amdgpus', not sure about nvidia's.

File: openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp (hunk 17 above)
Comment (carlobertolli) [reply openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp]: I changed the name to HSAXnackEnv because it does not refer to an OpenMP extension.

File: openmp/libomptarget/test/mapping/auto_zero_copy.cpp (hunk 16 above)
Comment (carlobertolli) [reply openmp/libomptarget/test/mapping/auto_zero_copy.cpp]: I added an info message so I can check that and remove libomptarget-debug.

**Comment by carlobertolli** [comment]
ping

**Review by saiislam — APPROVED** [review]
LGTM!
Thanks for the patch.

**Comment by carlobertolli** [comment]
As noted by a question during the review process, the test for zero-copy does not apply to dGPUs, including the gfx906 that runs the bbot. I will add an environment variable that allows users to turn on automatic zero-copy on any GPU, but that is not necessary if the system detected is an APU. That will make the test behave the same across different GPUs and enable users to test this even if they don't have any APU (some users are interested in this behavior for dGPUs).
